RFS/src/main/java/org/opensearch/migrations/bulkload/common/OpenSearchClientFactory.java

```diff
@@ -0,0 +1,45 @@
+package org.opensearch.migrations.bulkload.common;
+
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.opensearch.migrations.Version;
+import org.opensearch.migrations.VersionMatchers;
+import org.opensearch.migrations.bulkload.common.http.ConnectionContext;
+import org.opensearch.migrations.bulkload.version_es_6_8.OpenSearchClient_ES_6_8;
+import org.opensearch.migrations.bulkload.version_os_2_11.OpenSearchClient_OS_2_11;
+import org.opensearch.migrations.reindexer.FailedRequestsLogger;
+
+@RequiredArgsConstructor
+@Slf4j
+public class OpenSearchClientFactory {
+    // Version can be null, and if so, the "default" client, for OS_2_11 will be provided (matching the pre-factory behavior)
+    private final Version version;
```

Comment:
The version should be detectable - so we don't need to pass a parameter to this factory, here is an example https://github.com/opensearch-project/opensearch-migrations/pull/1226/commits/163d4829346dc228acc5519aa8257082b9380645

Reply:
@peternied I've been struggling with this point a bit, and I don't think that example is quite relevant. I agree that there's probably a better way to do this, but I don't think it's quite that easy, because you need a client to detect the version, but you can't create a client without a version

Reply:
So we're in a chicken-egg situation. We could have the client factory create a temporary client of the default (latest) version, check the version on the cluster, and then create a "real" client of the correct version, but I don't love that creating a Java class would then inherently put a load on the cluster--that seems like a pretty hidden side effect that could have consequences.

Reply:
@mikaylathompson I see your point, and I wish I was a bit more Java-savvy to propose a more elegant solution, but the general approach of having a "bootstrap" client that's our initial guess on the right one and finds the correct answer for us seems like the best overall direction.  The question is how to surface the potentially surprising side effect.  Maybe something like:

```java
var factory = new OpenSearchClientFactory()
factory.loadCorrectVersion()
var targetClient = factory.build() // Throws an exception if the loadCorrectVersion() method hasn't been invoked
```

Reply:
re: discussion with Peter, proposal is to update this to move the version-checking code (with the RestClient, not an OpenSearchClient) into the factory, and have an api a la:
```
var factory = new OpenSearchClientFactory();
var targetClient = factory.determineVersionAndCreate(connectionContext);
```